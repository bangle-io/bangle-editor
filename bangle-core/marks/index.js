export { Mark } from './mark';
export { Bold } from './bold';
export { Code } from './code';
export { Italic } from './italic';
export { Link } from './link';
export { Strike } from './strike';
export { Underline } from './underline';
