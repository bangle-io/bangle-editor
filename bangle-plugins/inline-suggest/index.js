export * as inlineSuggest from '../tooltip/suggestions-tooltip';
export * from './commands';
export * from './tooltip-controller';
export * from './trigger-input-rule';
export * as helpers from './helpers';
