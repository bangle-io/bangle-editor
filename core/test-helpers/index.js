import { NodeSelection } from 'prosemirror-state';

export * from './commands-helpers';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './schema-builders';
export * from './dispatch-paste-event';
export * from './create-event';
export * from './selection-helpers';

export const selectNodeAt = (view, pos) => {
  const tr = view.state.tr;
  view.dispatch(tr.setSelection(NodeSelection.create(tr.doc, pos)));
};
