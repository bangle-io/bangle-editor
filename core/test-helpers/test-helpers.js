import { NodeSelection } from 'prosemirror-state';
import { Fragment, Slice } from 'prosemirror-model';

export * from './commands-helpers';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './schema-builders';
export * from './dispatch-paste-event';
export * from './create-event';
export * from './selection-helpers';

export const selectNodeAt = (view, pos) => {
  const tr = view.state.tr;
  view.dispatch(tr.setSelection(NodeSelection.create(tr.doc, pos)));
};

/**
 *
 * @param {*} schema
 * @param {*} psxArray An array of psx nodes eg. [<para>hi</para>, <para>bye</para>]
 */
export const createPSXFragment = (schema, psxArray) => {
  return Fragment.fromArray(psxArray.map((r) => r(schema)));
};
