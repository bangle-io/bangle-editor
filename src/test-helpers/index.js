export * from './commands-helpers';
export * from './jest-helpers';
export * from './keyboard';
export * from './render-test-editor';
export * from './test-builders';
