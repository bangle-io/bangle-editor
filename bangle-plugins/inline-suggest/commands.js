import { isChromeWithSelectionBug } from 'bangle-core/index';
import { findFirstMarkPosition } from 'bangle-core/utils/pm-utils';
import { Fragment, Node } from 'prosemirror-model';
import { Selection } from 'prosemirror-state';
import { safeInsert } from 'prosemirror-utils';

export function removeTypeAheadMarkCmd(markType) {
  return (state, dispatch) => {
    const { from, to } = state.selection;

    const queryMark = findFirstMarkPosition(markType, state.doc, from - 1, to);

    const { start, end } = queryMark;
    if (
      start === -1 &&
      state.storedMarks &&
      markType.isInSet(state.storedMarks)
    ) {
      if (dispatch) {
        dispatch(state.tr.removeStoredMark(markType));
      }

      return true;
    }

    if (start === -1) {
      return false;
    }

    if (dispatch) {
      dispatch(
        state.tr
          .removeMark(start, end, markType)
          // stored marks are marks which will be carried forward to whatever
          // the user types next, like if current mark
          // is bold, new input continues being bold
          .removeStoredMark(markType)
          // This helps us avoid the case:
          // when a user deleted the trigger/ in '<trigger_mark>/something</trigger_mark>'
          // and then performs undo.
          // If we do not hide this from history, command z will bring
          // us in the state of `<trigger_mark>something<trigger_mark>` without the trigger `/`
          // and seeing this state `tooltipActivatePlugin` plugin will dispatch a new command removing
          // the mark, hence never allowing the user to command z.
          .setMeta('addToHistory', false),
      );
    }
    return true;
  };
}

export function selectItemCommand(maybeNode, markName) {
  return (state, dispatch, view) => {
    const { schema } = state;
    const markType = schema.marks[markName];
    const { selection } = state;
    const queryMark = findFirstMarkPosition(
      markType,
      state.doc,
      selection.from - 1,
      selection.to,
    );

    if (!queryMark || queryMark.start === -1) {
      return false;
    }

    const getTr = () => {
      const { start, end } = queryMark;
      let tr = state.tr
        .removeStoredMark(markType)
        .replaceWith(start, end, Fragment.empty);

      if (!maybeNode) {
        return tr;
      }

      const isInputFragment = maybeNode instanceof Fragment;

      let node;
      try {
        node =
          maybeNode instanceof Node || isInputFragment
            ? maybeNode
            : typeof maybeNode === 'string'
            ? state.schema.text(maybeNode)
            : Node.fromJSON(state.schema, maybeNode);
      } catch (e) {
        console.error(e);
        return tr;
      }

      if (node.isText) {
        tr = tr.replaceWith(start, start, node);
      } else if (node.isBlock) {
        tr = safeInsert(node)(tr);
      } else if (node.isInline || isInputFragment) {
        // throw new Error('nup');
        const fragment = isInputFragment
          ? node
          : Fragment.fromArray([node, state.schema.text(' ')]);

        tr = tr.replaceWith(start, start, fragment);
        // This problem affects Chrome v58+. See: https://github.com/ProseMirror/prosemirror/issues/710
        if (isChromeWithSelectionBug) {
          const selection = document.getSelection();
          if (selection) {
            selection.empty();
          }
        }

        // Placing cursor after node + space.
        tr = tr.setSelection(
          Selection.near(tr.doc.resolve(start + fragment.size)),
        );

        return tr;
      }

      return tr;
    };

    const tr = getTr();

    if (dispatch) {
      view.focus();
      dispatch(tr);
    }

    return true;
  };
}
