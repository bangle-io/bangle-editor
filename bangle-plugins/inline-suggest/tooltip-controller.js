import { Plugin } from 'prosemirror-state';
import { isMarkActiveInSelection } from 'bangle-core/utils/pm-utils';

import { doesQueryHaveTrigger } from './helpers';
import { removeTypeAheadMarkCmd } from './commands';
import {
  hideTooltip,
  showTooltip,
} from 'bangle-plugins/tooltip-placement/index';

const LOG = true;
let log = LOG
  ? console.log.bind(console, 'plugins/inline-tooltipActivatePlugin')
  : () => {};

/**
 * This plugin syncs the state between tooltip's show and trigger mark.
 * For example,
 * - if trigger mark is no longer active, it hides the tooltip.
 * - if it becomes active, it brings back the tooltip.
 * - if already active and mark position changed, it updates the tooltips position.
 * Because of this plugin, other plugins can derive their state just based on the tooltip's show state
 * @param {*} param0
 */
export function tooltipController({
  trigger,
  showTooltip,
  hideTooltip,
  markName,
}) {
  return new Plugin({
    view() {
      return {
        update: (view, lastState) => {
          const { state } = view;

          if (lastState === state || !state.selection.empty) {
            return;
          }
          const markType = state.schema.marks[markName];
          if (
            lastState.doc.eq(state.doc) &&
            state.selection.eq(lastState && lastState.selection) &&
            // This is a shorthand for checking if the stored mark  of `markType`
            // has changed within the last step. If it has we need to update the state
            isMarkStoredMark(state, markType) ===
              isMarkStoredMark(lastState, markType)
          ) {
            return;
          }

          const isMarkActive = isMarkActiveInSelection(markType)(state);

          // remove the mark if the user delete the trigger but remaining query
          // stayed. example `<mark>/hello</mark>` --(user deletes the /)-> `<mark>hello</mark>`
          if (isMarkActive && !doesQueryHaveTrigger(state, markType, trigger)) {
            removeTypeAheadMarkCmd(markType)(state, view.dispatch, view);
            return;
          }

          if (!isMarkActive) {
            hideTooltip(state, view.dispatch, view);
            return;
          }

          showTooltip(state, view.dispatch, view);
          return;
        },
      };
    },
  });
}

function isMarkStoredMark(state, markType) {
  return state && state.storedMarks && markType.isInSet(state.storedMarks);
}
