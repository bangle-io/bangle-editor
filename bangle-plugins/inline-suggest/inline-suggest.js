import { Mark } from 'bangle-core/marks/index';
import { findFirstMarkPosition, filter } from 'bangle-core/utils/pm-utils';

import {
  hideTooltip,
  showTooltip,
  tooltipPlacementPlugin,
} from '../tooltip-placement/index';
import { tooltipController } from './tooltip-controller';
import { triggerInputRule } from './trigger-input-rule';
import { getQueryText } from './helpers';
import { removeTypeAheadMarkCmd } from './commands';

const LOG = true;
let log = LOG ? console.log.bind(console, 'plugins/inline-suggest') : () => {};

export class InlineSuggest extends Mark {
  get name() {
    return 'inline_suggest_c' + this.options.trigger.charCodeAt(0);
  }

  get schema() {
    return {
      name: this.name,
      inclusive: true,
      group: 'triggerMarks',
      parseDOM: [{ tag: `span[data-${this.name}]` }],
      toDOM: (node) => {
        return [
          'span',
          {
            [`data-${this.name}`]: 'true',
            'data-trigger': node.attrs.trigger,
            'style': `color: #0052CC`,
          },
        ];
      },
      attrs: {
        trigger: { default: '' },
      },
    };
  }

  get markdown() {
    // This essentially removes the mark
    return {
      toMarkdown: {
        open: '',
        close: '',
        mixable: true,
      },
    };
  }

  get defaultOptions() {
    const rem = parseFloat(getComputedStyle(document.documentElement).fontSize);
    const { tooltipDOM, tooltipContent } = createTooltipDOM();
    tooltipContent.textContent = 'hello world';
    return {
      trigger: '/',
      tooltipDOM,
      placement: 'bottom-start',
      enterKeyName: 'Enter',
      arrowUpKeyName: 'ArrowUp',
      arrowDownKeyName: 'ArrowDown',
      escapeKeyName: 'Escape',
      fallbackPlacements: undefined,
      // Use another key to mimic enter behaviour for example, Tab for entering
      alternateEnterKeyName: undefined,
      getScrollContainerDOM: (view) => {
        return view.dom.parentElement;
      },
      tooltipOffset: () => {
        return [0, 0.4 * rem];
      },
      onUpdateTooltip: (state, dispatch, view) => {},
      // No need to call removeTypeAheadMarkCmd on destroy
      onHideTooltip: (state, dispatch, view) => {},
      onEnter: (state, dispatch, view) => {
        return removeTypeAheadMarkCmd(this.getMarkType(state))(
          state,
          dispatch,
          view,
        );
      },
      onArrowDown: (state, dispatch, view) => {
        return true;
      },
      onArrowUp: (state, dispatch, view) => {
        return true;
      },
      onEscape: (state, dispatch, view) => {
        return removeTypeAheadMarkCmd(this.getMarkType(state))(
          state,
          dispatch,
          view,
        );
      },
    };
  }

  inputRules({ schema }) {
    return [triggerInputRule(schema, this.name, this.options.trigger)];
  }

  /**
   *
   * @param {*} maybeNode Node | string | Fragment
   * @param {*} opts
   */

  get plugins() {
    const plugin = tooltipPlacementPlugin({
      pluginName: this.name + 'Tooltip',
      placement: this.options.placement,
      fallbackPlacements: this.options.fallbackPlacements,
      tooltipOffset: this.options.tooltipOffset,
      tooltipDOM: this.options.tooltipDOM,
      getScrollContainerDOM: this.options.getScrollContainerDOM,
      getReferenceElement: triggerReferenceElement(
        (state) => this.getMarkType(state),
        (state, markType) => {
          const { selection } = state;
          return findFirstMarkPosition(
            markType,
            state.doc,
            selection.from - 1,
            selection.to,
          );
        },
      ),
      showTooltipArrow: false,
      onUpdateTooltip: this.options.onUpdateTooltip,
      onHideTooltip: this.options.onHideTooltip,
    });

    this._tooltipPlugin = plugin;

    const { trigger } = this.options;
    const markName = this.name;

    return [
      plugin,
      tooltipController({
        trigger,
        markName,
        showTooltip: showTooltip(plugin),
        hideTooltip: hideTooltip(plugin),
      }),
    ];
  }

  keys() {
    const isActiveCheck = (state) => this.isTooltipActive(state);

    const result = {
      [this.options.enterKeyName]: filter(isActiveCheck, this.options.onEnter),
      [this.options.arrowUpKeyName]: filter(
        isActiveCheck,
        this.options.onArrowUp,
      ),
      [this.options.arrowDownKeyName]: filter(
        isActiveCheck,
        this.options.onArrowDown,
      ),
      [this.options.escapeKeyName]: filter(
        isActiveCheck,
        this.options.onEscape,
      ),
    };
    if (this.options.alternateEnterKeyName) {
      result[this.options.alternateEnterKeyName] =
        result[this.options.enterKeyName];
    }
    return result;
  }

  /**
   * Public API
   */

  getMarkName() {
    return this.name;
  }

  getMarkType(state) {
    return state.schema.marks[this.name];
  }

  getQueryText(state) {
    const markType = this.getMarkType(state);
    return getQueryText(state, markType, this.options.trigger);
  }

  isTooltipActive(state) {
    return this._tooltipPlugin && this._tooltipPlugin.getState(state)?.show;
  }

  getTooltipPlugin() {
    return this._tooltipPlugin;
  }
}

export function createTooltipDOM(className = 'bangle-tooltip') {
  const tooltipDOM = document.createElement('div');
  tooltipDOM.className = className;
  tooltipDOM.setAttribute('role', 'tooltip');
  const tooltipContent = document.createElement('div');
  tooltipContent.className = className + '-content';
  tooltipDOM.appendChild(tooltipContent);
  return { tooltipDOM, tooltipContent };
}

export function triggerReferenceElement(getMarkType, getActiveMarkPos) {
  return (view, tooltipDOM, scrollContainerDOM) => {
    const markType = getMarkType(view.state);
    return {
      getBoundingClientRect: () => {
        let state = view.state;
        const markPos = getActiveMarkPos(state, markType);
        // add by + so that we get the position right after trigger
        const startPos = markPos.start > -1 ? markPos.start + 1 : 0;
        const start = view.coordsAtPos(startPos);

        // if the query spanned two lines, we want to show the tooltip based on the end pos
        // so that it doesn't hide the query
        const end = view.coordsAtPos(markPos.end > -1 ? markPos.end : startPos);

        let { left, right } = start;
        let { top, bottom } = end;

        let z = {
          width: right - left,
          height: bottom - top,
          top: top,
          right: left,
          bottom: bottom,
          left: left,
        };

        return z;
      },
    };
  };
}
