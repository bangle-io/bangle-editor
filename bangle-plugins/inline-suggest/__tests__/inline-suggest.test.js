/**
 * @jest-environment jsdom
 */

/** @jsx psx */

import {
  psx,
  typeText,
  renderTestEditor,
  sendKeyToPm,
} from 'bangle-core/test-helpers/index';

import {
  Blockquote,
  BulletList,
  CodeBlock,
  HardBreak,
  Heading,
  ListItem,
  OrderedList,
  TodoItem,
  TodoList,
} from 'bangle-core/nodes/index';
import { InlineSuggest } from '../inline-suggest';
import { typeChar } from 'bangle-core/test-helpers/index';
import { Selection } from 'prosemirror-state';
import { Italic, Underline } from 'bangle-core/marks/index';

// We are using char code to differentiate between different schema
// 47 is char code for '/'
const triggerMarkSlash = (content) => (
  <inline_suggest_c47 trigger="/">{content}</inline_suggest_c47>
);

describe('inline suggest basic show and hide', () => {
  let suggestionExtension, testEditor;
  const getTooltipState = (state) => {
    return suggestionExtension.getTooltipPlugin().getState(state);
  };

  beforeEach(async () => {
    suggestionExtension = new InlineSuggest({
      trigger: '/',
    });

    const extensions = [
      new BulletList(),
      new ListItem(),
      new OrderedList(),
      new HardBreak(),
      new Heading(),
      new Underline(),
      new TodoList(),
      new TodoItem(),
      new Blockquote(),
      new CodeBlock(),
      new Italic(),
      suggestionExtension,
    ];
    testEditor = renderTestEditor({ extensions });
  });

  test('when no trigger', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>foo[]bar</para>
      </doc>,
    );

    expect(editor.view.dom.parentNode).toMatchSnapshot();
  });

  test.each([
    <doc>
      <para>[]</para>
    </doc>,
    <doc>
      <para>hello []</para>
    </doc>,
    <doc>
      <para>hello</para>
      <para>hello</para>
      <para>hello []</para>
    </doc>,
    <doc>
      <ul>
        <li>
          <para>hello []</para>
        </li>
      </ul>
    </doc>,
    <doc>
      <todoList>
        <todoItem>
          <para>hello []</para>
        </todoItem>
      </todoList>
    </doc>,
    <doc>
      <ol>
        <li>
          <para>hello []</para>
        </li>
      </ol>
    </doc>,
    <doc>
      <blockquote>
        <heading>hello []</heading>
      </blockquote>
    </doc>,
    <doc>
      <blockquote>
        <heading>hello</heading>
      </blockquote>
      <para>[]</para>
    </doc>,
  ])('Case %# different parent block', async (input) => {
    const { editor } = await testEditor(input);

    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(false);
    typeChar(editor.view, '/');
    typeText(editor.view, 'check');
    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(true);
    expect(suggestionExtension.getQueryText(editor.state)).toBe('check');
    expect(editor.state.doc.toString()).toMatchSnapshot();

    expect(getTooltipState(editor.state)).toEqual({
      show: true,
    });
  });

  test('Not allowed in codeblock', async () => {
    const { editor } = await testEditor(
      <doc>
        <ul>
          <li>
            <para>top</para>
          </li>
        </ul>
        <codeBlock>hello []</codeBlock>
      </doc>,
    );

    typeChar(editor.view, '/');
    typeText(editor.view, 'check');

    expect(editor.state).toEqualDocAndSelection(
      <doc>
        <ul>
          <li>
            <para>top</para>
          </li>
        </ul>
        <codeBlock>hello /check[]</codeBlock>
      </doc>,
    );

    expect(getTooltipState(editor.state)).toEqual({
      show: false,
    });
  });

  test('trigger on an empty doc', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>[]</para>
      </doc>,
    );

    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(false);

    typeChar(editor.view, '/');
    typeText(editor.view, 'check');
    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(true);
    expect(suggestionExtension.getQueryText(editor.state)).toBe('check');

    expect(editor.state).toEqualDocAndSelection(
      <doc>
        <para>{triggerMarkSlash('/check')}[]</para>
      </doc>,
    );

    expect(getTooltipState(editor.state)).toEqual({
      show: true,
    });
  });

  test('types a trigger', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>foobar []</para>
      </doc>,
    );

    typeChar(editor.view, '/');
    typeText(editor.view, 'check');
    expect(suggestionExtension.getQueryText(editor.state)).toBe('check');

    expect(editor.state).toEqualDocAndSelection(
      <doc>
        <para>foobar {triggerMarkSlash('/check')}[]</para>
      </doc>,
    );

    expect(getTooltipState(editor.state)).toEqual({
      show: true,
    });

    expect(editor.view.dom.parentNode).toMatchSnapshot();
  });

  test('Selection going to other location hides the tooltip', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(editor.view, '/');
    typeText(editor.view, '');
    const { state } = editor.view;

    // Outside the mark
    editor.view.dispatch(state.tr.setSelection(Selection.atEnd(state.doc)));

    expect(getTooltipState(editor.state)).toEqual({
      show: false,
    });

    let tr = state.tr;

    // Inside the mark
    editor.view.dispatch(tr.setSelection(Selection.near(tr.doc.resolve(2))));
    expect(getTooltipState(editor.state)).toEqual({
      show: true,
    });
    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(true);
  });

  test('Selection going to other location hides the tooltip', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(editor.view, '/');
    typeText(editor.view, '');
    const { state } = editor.view;

    // Outside the mark
    editor.view.dispatch(state.tr.setSelection(Selection.atEnd(state.doc)));

    expect(getTooltipState(editor.state)).toEqual({
      show: false,
    });

    let tr = state.tr;

    // Inside the mark
    editor.view.dispatch(tr.setSelection(Selection.near(tr.doc.resolve(2))));
    expect(getTooltipState(editor.state)).toEqual({
      show: true,
    });
    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(true);
  });

  test('Query at end of document', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(editor.view, '/');
    typeText(editor.view, 'first');
    const view = editor.view;

    // Outside the mark
    editor.view.dispatch(
      view.state.tr.setSelection(Selection.atEnd(view.state.doc)),
    );

    typeText(editor.view, ' ');
    typeChar(editor.view, '/');
    typeText(editor.view, 'second');

    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
    expect(suggestionExtension.getQueryText(view.state)).toBe('second');

    let tr = view.state.tr;

    // Inside the mark
    editor.view.dispatch(tr.setSelection(Selection.near(tr.doc.resolve(7))));
    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
    expect(suggestionExtension.getQueryText(view.state)).toBe('first');
  });

  test('Really long query', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    const longText =
      'firstfirstfirstfirst firstfirstfirstfirstfirst firstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirst firstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirstfirst';
    typeChar(editor.view, '/');
    typeText(editor.view, longText);

    const view = editor.view;
    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
    expect(suggestionExtension.getQueryText(view.state)).toBe(longText);
  });

  test('Typing query slowly', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>[] hello</para>
      </doc>,
    );

    typeChar(editor.view, '/');
    typeText(editor.view, 'c');
    typeText(editor.view, 'h');
    typeText(editor.view, 'e');
    typeText(editor.view, 'c');
    typeText(editor.view, 'k');

    const view = editor.view;
    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
    expect(suggestionExtension.getQueryText(view.state)).toBe('check');
  });

  test.skip('When forward deleting the mark the tooltip hides', async () => {
    const { editor } = await testEditor(
      <doc>
        <para>hello[]</para>
      </doc>,
    );

    const { view } = editor;
    // typeChar(view, '/');
    // typeText(view, 'check');
    // expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);

    // view.dispatch(view.state.tr.delete(1, 2));

    // editor.view.dispatch(
    //   view.state.tr.setSelection(Selection.atStart(view.state.doc)),
    // );

    sendKeyToPm(view, 'Z');
    // sendKeyToPm(view, 'Backspace');
    // sendKeyToPm(view, 'Backspace');

    expect(editor.state).toEqualDocAndSelection(
      <doc>
        <para>hello</para>
      </doc>,
    );
    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
  });
});

describe('keybindings test', () => {
  let suggestionExtension, testEditor;

  const opts = {
    trigger: '/',
    placement: 'bottom-start',
    enterKeyName: 'Enter',
    arrowUpKeyName: 'ArrowUp',
    arrowDownKeyName: 'ArrowDown',
    escapeKeyName: 'Escape',
    // Use another key to mimic enter behaviour for example, Tab for entering
    alternateEnterKeyName: undefined,

    onUpdateTooltip: jest.fn(() => true),
    onHideTooltip: jest.fn(() => true),
    onEnter: jest.fn(() => true),
    onArrowDown: jest.fn(() => true),
    onArrowUp: jest.fn(() => true),
    onEscape: jest.fn(() => true),
  };

  test('calls on* callbacks correctly', async () => {
    suggestionExtension = new InlineSuggest(opts);

    const extensions = [
      new BulletList(),
      new ListItem(),
      new OrderedList(),
      new HardBreak(),
      new Heading(),
      new Underline(),
      new TodoList(),
      new TodoItem(),
      new Blockquote(),
      new CodeBlock(),
      new Italic(),
      suggestionExtension,
    ];
    testEditor = renderTestEditor({ extensions });
    const { editor } = await testEditor(
      <doc>
        <para>[] foobar</para>
      </doc>,
    );

    const view = editor.view;
    typeChar(view, '/');
    typeText(view, 'check');
    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
    expect(suggestionExtension.getQueryText(editor.state)).toBe('check');

    expect(editor.state).toEqualDocAndSelection(
      <doc>
        <para>{triggerMarkSlash('/check')}[] foobar</para>
      </doc>,
    );
    // since it is called many times
    expect(opts.onUpdateTooltip).toBeCalled();
    expect(opts.onHideTooltip).toBeCalledTimes(0);
    expect(suggestionExtension.isTooltipActive(editor.state)).toBe(true);

    sendKeyToPm(view, 'Escape');
    expect(opts.onEscape).toBeCalledTimes(1);
    sendKeyToPm(view, 'Enter');
    expect(opts.onEnter).toBeCalledTimes(1);
    sendKeyToPm(view, 'ArrowUp');
    expect(opts.onArrowUp).toBeCalledTimes(1);
    sendKeyToPm(view, 'ArrowDown');
    expect(opts.onArrowDown).toBeCalledTimes(1);

    expect(suggestionExtension.isTooltipActive(view.state)).toBe(true);
    expect(suggestionExtension.getQueryText(editor.state)).toBe('check');

    // Outside the mark
    view.dispatch(view.state.tr.setSelection(Selection.atEnd(view.state.doc)));
    // TODO for some mysterious reason this is called twice with the same state
    expect(opts.onHideTooltip).toBeCalledTimes(1);
  });
});

test.todo(
  'Check with differen trigger inline-suggest and that they dont interfere with each other',
);
