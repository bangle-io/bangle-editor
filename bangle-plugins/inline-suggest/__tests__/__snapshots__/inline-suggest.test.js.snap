// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inline suggest basic show and hide Case 0 different parent block 1`] = `"doc(paragraph(inline_suggest_c47(\\"/check\\")))"`;

exports[`inline suggest basic show and hide Case 1 different parent block 1`] = `"doc(paragraph(\\"hello \\", inline_suggest_c47(\\"/check\\")))"`;

exports[`inline suggest basic show and hide Case 2 different parent block 1`] = `"doc(paragraph(\\"hello\\"), paragraph(\\"hello\\"), paragraph(\\"hello \\", inline_suggest_c47(\\"/check\\")))"`;

exports[`inline suggest basic show and hide Case 3 different parent block 1`] = `"doc(bullet_list(list_item(paragraph(\\"hello \\", inline_suggest_c47(\\"/check\\")))))"`;

exports[`inline suggest basic show and hide Case 4 different parent block 1`] = `"doc(todo_list(todo_item(paragraph(\\"hello \\", inline_suggest_c47(\\"/check\\")))))"`;

exports[`inline suggest basic show and hide Case 5 different parent block 1`] = `"doc(ordered_list(list_item(paragraph(\\"hello \\", inline_suggest_c47(\\"/check\\")))))"`;

exports[`inline suggest basic show and hide Case 6 different parent block 1`] = `"doc(blockquote(heading(\\"hello \\", inline_suggest_c47(\\"/check\\"))))"`;

exports[`inline suggest basic show and hide Case 7 different parent block 1`] = `"doc(blockquote(heading(\\"hello\\")), paragraph(inline_suggest_c47(\\"/check\\")))"`;

exports[`inline suggest basic show and hide types a trigger 1`] = `
<div
  id="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
    tabindex="0"
  >
    <p>
      foobar 
      <span
        data-inline_suggest_c47="true"
        data-trigger="/"
        style="color: #0052CC"
      >
        /check
      </span>
    </p>
  </div>
  <div
    class="bangle-tooltip"
    data-show=""
    data-tooltip-name="inline_suggest_c47Tooltip"
    role="tooltip"
    style="position: absolute; left: 0px; top: 0px; margin: 0px;"
  >
    <div
      class="bangle-tooltip-content"
    >
      hello world
    </div>
  </div>
</div>
`;

exports[`inline suggest basic show and hide when no trigger 1`] = `
<div
  id="test-editor"
>
  <div
    class="ProseMirror bangle-editor content"
    contenteditable="true"
    tabindex="0"
  >
    <p>
      foobar
    </p>
  </div>
  <div
    class="bangle-tooltip"
    data-tooltip-name="inline_suggest_c47Tooltip"
    role="tooltip"
  >
    <div
      class="bangle-tooltip-content"
    >
      hello world
    </div>
  </div>
</div>
`;
